### PR TITLE
Adding step to ensure no replace directives

### PR DESCRIPTION
**Pull request by seankane-msft**
#16046 

- [ ] The purpose of this PR is explained in this or a referenced issue.
- [ ] The PR does not update generated files.
   - These files are managed by the codegen framework at [Azure/autorest.go][].
- [ ] Tests are included and/or updated for code changes.
- [ ] Updates to [CHANGELOG.md][] are included.
- [ ] MIT license headers are included in each file.

[Azure/autorest.go]: https://github.com/Azure/autorest.go
[CHANGELOG.md]: https://github.com/Azure/azure-sdk-for-go/blob/main/CHANGELOG.md

### CODE REVIEW

File: eng/scripts/validate_go_mod.ps1
Comment (jhendrixMSFT) [inline eng/scripts/validate_go_mod.ps1:23]: Nit: missing new-line

File: eng/pipelines/templates/steps/analyze.yml
Comment (jhendrixMSFT) [inline eng/pipelines/templates/steps/analyze.yml:40]: 'Require no replace directives in go.mod file'

**Review by jhendrixMSFT — APPROVED** [review]
Few minor things.  Also, this will require us to get an updated azidentity released for any modules that require it as part of examples/tests.

**Comment by chlowell** [comment]
Can we run this only in release pipelines? I imagine we'll sometimes want to develop against unreleased azcore features, and  that's okay--we really just want to prevent shipping in that state.

**Comment by seankane-msft** [comment]
> Can we run this only in release pipelines? I imagine we'll sometimes want to develop against unreleased azcore features, and that's okay--we really just want to prevent shipping in that state.

Done, now waiting on @benbp approval before merging. Not as familiar with the release portions.

File: eng/pipelines/templates/jobs/archetype-go-release.yml
Comment (RickWinter) [inline eng/pipelines/templates/jobs/archetype-go-release.yml:33]: Nit
```suggestion
                    displayName: 'Verify go.mod file does NOT contain replace directives'
```

**Review by RickWinter — APPROVED** [review]

File: eng/pipelines/templates/jobs/archetype-go-release.yml
Comment (benbp) [reply eng/pipelines/templates/jobs/archetype-go-release.yml:33]: I would also suggest below because it will probably show up in the truncated display name UI as "Verify go.mod file"

```suggestion
                    displayName: 'Verify no replace directives in go.mod file'
```

File: eng/scripts/validate_go_mod.ps1
Comment (benbp) [inline eng/scripts/validate_go_mod.ps1:10]:
```suggestion
    Write-Host "Could not find a go.mod file in the directory $(pwd)"
```

File: eng/scripts/validate_go_mod.ps1
Comment (benbp) [inline eng/scripts/validate_go_mod.ps1:15]:
```suggestion
    $patternMatches = Get-Content $goMod.FullName | Select-String -Pattern " replace "
```

File: eng/scripts/validate_go_mod.ps1
Comment (benbp) [inline eng/scripts/validate_go_mod.ps1:19]: I would log here and set a variable like `$hasError` to true, then exit after the loop. That way you validate all the go.mod files up front (makes it easier for the pipeline tester).

File: eng/scripts/validate_go_mod.ps1
Comment (seankane-msft) [reply eng/scripts/validate_go_mod.ps1:15]: Why the space beforehand? The replace directive should start at the beginning of a new line

File: eng/scripts/validate_go_mod.ps1
Comment (benbp) [inline eng/scripts/validate_go_mod.ps1:19]: Nit (two options):

```suggestion
    $patternMatches = Get-Content $goMod.FullName | Select-String -Pattern "replace "
    if ($patternMatches.Length -ne 0) {
        Write-Host "Found a replace directive in go.mod file at $($goMod.FullName)"
```

```suggestion
    $name = $goMod.FullName
    $patternMatches = Get-Content $name | Select-String -Pattern "replace "
    if ($patternMatches.Length -ne 0) {
        Write-Host "Found a replace directive in go.mod file at $name"
```

**Review by benbp — APPROVED** [review]

**Pull request merged**
